src/tasks/SingularityHandler.cpp

```diff
@@ -0,0 +1,346 @@
+        _singularity_history.push_back(TYPE_2_SINGULARITY);
+        _type_2_counter++;
+    }
+
+    // pop oldest if greater than buffer size
+    if (_singularity_history.size() > _buffer_size) {
+        if (_singularity_history.front() == TYPE_1_SINGULARITY) {
+            _type_1_counter--;
+        } else if (_singularity_history.front() == TYPE_2_SINGULARITY) {
+            _type_2_counter--;
+        }
+        _singularity_history.pop_front();
+    }
+
+}
+
+VectorXd SingularityHandler::computeTorques(const VectorXd& unit_mass_force, const VectorXd& force_related_terms) {
+    if (_verbose) {
+        if (_singularity_types.size() != 0) {
+            for (auto type : _singularity_types) {
+                std::cout << "Singularity: " << singularity_labels[type] << " | ";
+            }
+            std::cout << "\n---\n";
+        }
+    }
+
+    if (_singularity_types.size() == 0) {
+        return _projected_jacobian_ns.transpose() * (_Lambda_ns_modified * _task_range_ns.transpose() * unit_mass_force + \
+                    _task_range_ns.transpose() * force_related_terms);
+    } else {
+        VectorXd tau_ns = VectorXd::Zero(_dof);
+
+        // compute non-singular torques 
+        if (_task_range_ns.norm() == 0) {
+            return tau_ns;
+        } else {
+            tau_ns = _projected_jacobian_ns.transpose() * (_Lambda_ns_modified * _task_range_ns.transpose() * unit_mass_force + \
+                        _task_range_ns.transpose() * force_related_terms);
+            if (!_enforce_joint_strategy) {
+                return tau_ns;
+            }
+        } 
+
+        // handle singularity type based on which one has more counts  
+        if (_type_1_counter > _type_2_counter || _enforce_type_1_strategy) {
+            // joint holding to entering joint conditions  
+            VectorXd unit_torques = - _kp_type_1 * (_robot->q() - _q_prior) - _kv_type_1 * _robot->dq();  
+            _joint_strategy_torques = _posture_projected_jacobian.transpose() * _Lambda_joint_s_modified * _joint_task_range_s.transpose() * unit_torques;
+        } else {
+            // apply open-loop torque proportional to dot(unit mass force, singular direction)
+            // zero torque achieved when singular direction is orthogonal to the desired unit mass force direction
+            // the direction is reversed if the joint is approaching a joint limit 
+            for (int i = 0; i < _joint_task_range_s.rows(); ++i) {
+                if (_joint_task_range_s(i, 0) != 0) {
+                    if (std::abs(_robot->q()(i) - _q_upper(i)) < 5 * M_PI / 180) {
+                        _type_2_direction(i) = - 1;
+                    } else if (std::abs(_robot->q()(i) - _q_lower(i)) < 5 * M_PI / 180) {
+                        _type_2_direction(i) = 1;
+                    } 
+                }
+            }
+            double fTd = ((unit_mass_force.normalized() + force_related_terms.normalized()).normalized()).dot(_task_range_s.col(0));
```

Comment:
probably more correct to sum the unit mass force and the force related term, and only normalize their sum